Userland/Services/NotificationServer/NotificationWindow.cpp

```diff
@@ -92,36 +108,59 @@ NotificationWindow::NotificationWindow(const String& text, const String& title,
     widget.layout()->set_spacing(6);
 
     if (icon.is_valid()) {
-        auto& image = widget.add<GUI::ImageWidget>();
-        image.set_bitmap(icon.bitmap());
+        m_image = &widget.add<GUI::ImageWidget>();
+        m_image->set_bitmap(icon.bitmap());
     }
 
     auto& left_container = widget.add<GUI::Widget>();
     left_container.set_layout<GUI::VerticalBoxLayout>();
 
-    auto& title_label = left_container.add<GUI::Label>(title);
-    title_label.set_font(Gfx::FontDatabase::default_bold_font());
-    title_label.set_text_alignment(Gfx::TextAlignment::CenterLeft);
-    auto& text_label = left_container.add<GUI::Label>(text);
-    text_label.set_text_alignment(Gfx::TextAlignment::CenterLeft);
+    m_title_label = &left_container.add<GUI::Label>(title);
+    m_title_label->set_font(Gfx::FontDatabase::default_bold_font());
+    m_title_label->set_text_alignment(Gfx::TextAlignment::CenterLeft);
+    m_text_label = &left_container.add<GUI::Label>(text);
+    m_text_label->set_text_alignment(Gfx::TextAlignment::CenterLeft);
 
     widget.set_tooltip(text);
-    title_label.set_tooltip(text);
-    text_label.set_tooltip(text);
+    m_title_label->set_tooltip(text);
+    m_text_label->set_tooltip(text);
 
     auto& right_container = widget.add<GUI::Widget>();
     right_container.set_fixed_width(36);
     right_container.set_layout<GUI::HorizontalBoxLayout>();
 
-    on_close_request = [this] {
-        s_windows.remove_first_matching([this](auto& entry) { return entry == this; });
+    on_close_request = [&] {
+        s_windows.remove(m_window_id);
         update_notification_window_locations();
         return CloseRequestDecision::Close;
     };
 }
 
 NotificationWindow::~NotificationWindow()
 {
+    s_windows.remove(m_window_id);
+}
+
+RefPtr<NotificationWindow> NotificationWindow::get_window_by_id(i64 id)
+{
+    VERIFY(id >= 0);
+    auto window = s_windows.get(id);
+    return window.value_or(nullptr);
+}
+
+void NotificationWindow::set_text(const String& text)
+{
+    m_text_label->set_text(text);
+}
+
+void NotificationWindow::set_title_widget(const String& text)
+{
+    m_title_label->set_text(text);
+}
+
+void NotificationWindow::set_image(const Gfx::ShareableBitmap& icon)
+{
+    m_image->set_bitmap(icon.bitmap());
```

Comment:
This will crash if the original notification didn't have an image (see `if (icon.is_valid()) {`), because in that case `m_image` is `nullptr`.